internal/provider/resource_secret.go

```diff
@@ -0,0 +1,281 @@
+				ElementType: types.StringType,
+				Required:    true,
+				Sensitive:   true,
+			},
+			"secret_id": schema.StringAttribute{
+				Description: "The ID of the secret.",
+				Computed:    true,
+			},
+			"info": schema.StringAttribute{
+				Description: "The description of the secret.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func (s *secretResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	s.client = client
+	// Create the local logging subsystem here, using the TF context when creating it.
+	s.subCtx = tflog.NewSubsystem(ctx, LogResourceSecret)
+}
+
+func (s *secretResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret", "create")
+		return
+	}
+
+	var plan secretResourceModel
+
+	// Read Terraform plan into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var secretValue map[string]string
+	resp.Diagnostics.Append(plan.Value.ElementsAs(ctx, &secretValue, false)...)
+
+	createSecretOutput, err := s.client.Secrets.CreateSecret(&juju.CreateSecretInput{
+		ModelName: plan.Model.ValueString(),
+		Name:      plan.Name.ValueString(),
+		Value:     secretValue,
+		Info:      plan.Info.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to add secret, got error: %s", err))
+		return
+	}
+	s.trace(fmt.Sprintf("add secret resource %q", plan.SecretId))
+
+	plan.SecretId = types.StringValue(createSecretOutput.SecretId)
+
+	// Save plan into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+
+	s.trace(fmt.Sprintf("create secret resource %q", plan.SecretId))
+}
+
+func StringPtr(s string) *string {
+	return &s
+}
+
+func (s *secretResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret", "read")
+		return
+	}
+
+	var state secretResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	readSecretOutput, err := s.client.Secrets.ReadSecret(&juju.ReadSecretInput{
+		SecretId:  state.SecretId.ValueString(),
+		ModelName: state.Model.ValueString(),
+		Name:      StringPtr(state.Name.ValueString()),
```

Comment:
The SecretID may be nil as well on the first read or on import.

 Use `state.Name.ValueStringPointer()`. `StringPtr(state.Name.ValueString())` will always return a pointer which sometimes may point to an empty string.